File: laboratories/ppo/exercises/lab03/lab03.py
import random


class Player:
    name = None
    __health_points = 100

    def attack(self, player):
        hit_points = random.randint(1, 3) * 10
        player.take_hit(hit_points)

        return hit_points

    def take_hit(self, hit_points):
        self.__health_points = self.__health_points - hit_points

    def is_dead(self):
        return self.__health_points <= 0


def get_random_player(players):
    return random.choice(players)


players = []

luke = Player()
luke.name = "Luke Skywalker"
players.append(luke)

vader = Player()
vader.name = "Darth Vader"
players.append(vader)

round = 1
ongoing = True

while ongoing:
    attacker = get_random_player(players)
    target = get_random_player(players)
    hit_points = attacker.attack(target)

    print(
        "Round "
        + str(round)
        + ": "
        + attacker.name
        + " attacked "
        + target.name
        + " with "
        + str(hit_points)
        + " points."
    )

    for player in players:
        if player.is_dead():
            print("Player " + player.name + " died.")
            ongoing = False

    round = round + 1

File: laboratories/ppo/exercises/lab03/lab03.java
import java.util.ArrayList;
import java.util.Random;

class Player {
  public String name;
  protected int healthPoints = 100;

  public int attack(Player player) {
    int hitPoints = (new Random().nextInt(3) + 1) * 10;
    player.takeHit(hitPoints);

    return hitPoints;
  }

  public void takeHit(int hitPoints) {
    this.healthPoints = this.healthPoints - hitPoints;
  }

  public boolean isDead() {
    return this.healthPoints <= 0;
  }
}

public class lab03 {
  public static void main(String[] args) {
    ArrayList < Player > players = new ArrayList < Player > ();

    Player luke = new Player();
    luke.name = "Luke Skywalker";
    players.add(luke);

    Player vader = new Player();
    vader.name = "Darth Vader";
    players.add(vader);

    int round = 1;
    boolean ongoing = true;

    while (ongoing) {
      Player attacker = getRandomPlayer(players);
      Player target = getRandomPlayer(players);
      int hitPoints = attacker.attack(target);

      System.out.println("Round " + round + ": " + attacker.name + " attacked " + target.name + " with " + hitPoints + " points.");

      for (Player p: players) {
        if (p.isDead()) {
          System.out.println("Player " + p.name + " died.");
          ongoing = false;
        }
      }

      round++;
    }
  }

  static Player getRandomPlayer(ArrayList < Player > players) {
    return players.get(new Random().nextInt(players.size()));
  }
}
